README.md
<img src="https://img.shields.io/npm/v/webperf-comparison" /> <img src="https://img.shields.io/node/v/webperf-comparison" />

# Web performance comparison tool

## Purpose

- Run lighthouse in a headless mode in an automated fashion
- [Able to perform multiple runs in order to get a better aggregate result](https://github.com/GoogleChrome/lighthouse/blob/HEAD/docs/variability.md)
- Perform a comparison between two hosted experiences
- Grok Lighthouse's insanely large JSON response

## Pre-requisites

- Node 14+, I haven't tested on any previous versions, so ultimately it could work prior, but I haven't validated it. YMMV
- Chrome installed -- Necessary for running lighthouse!

## Target audience

**Developers** - I didn't implement any error logging because there aren't many recoverable steps. If errors arise, I feel the errors in the dependencies themselves provide enough context as to what went wrong. Hence, my target audience is one who can read & understand unhandled promise rejections.

## Starting

### by way of cloning this project

- Install the packages, using `npm i`
- Start the prompts using `npm start`

### via the published module

- Run the command `npx webperf-comparison`

1. Answer all the prompts (note: light gray values represent defaults, pressing enter accepts the default, and any keystroke erases the default)
1. Let lighthouse do its magic, the progress bar should keep you up to date on status
1. See the output table

### Previous Use

If you want to re-evaluate the previous run, which leverages the `results.json` written to disk, you can use the command: `webperf-comparison --usePrevious` or, if cloned, `npm run start:previous`

## Want to run your own Chromium? Use the CHROME_PATH environment argument

`chrome-launcher` allows you to define the CHROME_PATH for the launcher. For instance, if you wanted to run Brave Browser, you could do `CHROME_PATH="C:\Program Files (x86)\BraveSoftware\Brave-Browser\Application\brave.exe" npm start`, and huzzah, it works!

## Implementation

- [prompts](https://www.npmjs.com/package/prompts) - For asking the leading questions
- [lighthouse](https://www.npmjs.com/package/lighthouse) - For analyzing web performance

Using prompts, we gather the URL of the host(s) to run against. Additionally, we ask a series of path segments to test against for each experience and how many times to run the test. Many of these values can be set by a [.webperfrc](#webperfc) for convenience.

## <a name="webperfc"></a> .webperfrc

The configuration file can be defined in a number of ways, as per the [rc module documentation](https://www.npmjs.com/package/rc). The simplest way is to provide the file by way of `--config=.webperfrc`.

| Property      | Description                                                                                |
| ------------- | ------------------------------------------------------------------------------------------ |
| runs          | The number of times to run lighthouse against each URL/route combination                   |
| hosts        | A comma delimited list of hosts to test against |
| routes        | A comma delimited list of paths to test per URL (i.e. /, /test)                            |
| loadSite      | Whether to load the results details site upon finish, leaving just the command line output |
| chromePath    | Path to Chromium install, optional                                                         |
| network | Network throttling option, see [Network throttling](#network_throttling) |
| cpu | CPU throttling option, see [CPU throttling](#cpu_throttling) |

## <a name="network_throttling"></a> Network throttling (case insensitive)

Lighthouse allows network throttling to be simulated. I had looked at [throttle](https://www.sitespeed.io/documentation/throttle/) to enable this at the hardware level, but it would require sudo permissions and only works on Mac/Linux. Instead, I opted for simulated network throttling which is built into lighthouse itself.

| Option | Round trip time (ms) | Upload speed (Kbps) | Download speed (Kbps) |
| ------ | -------------------- | ------------------- | --------------------- |
| 3gslow | 200                  | 400                 | 400                   |
| 3g     | 150                  | 768                 | 1600                  |
| 3gfast | 75                   | 768                 | 1600                  |
| 4g     | 85                   | 9000                | 9000                  |
| cable  | 14                   | 1000                | 5000                  |

## <a name="cpu_throttling"></a> CPU throttling (case insensitive)

Additional to network throttling, Lighthouse allows CPU throttling which helps mimic processing capabilities of mobile phones vs the device running this module. This is important for testing things like LCP/FID/Blocking time, etc.

| Option            | CPU modifier, 1/modifier | Description |
| ----------------- | ------------------------ | ----------- |
| Slow mobile phone | 6                        | Slow mobile |
| Mobile phone      | 4                        | Mobile      |
| Desktop           | 1                        | Your device |
## Site development

Want to help change what the site output is? That's great! Thanks for the help, in advance. You will have to clone this repository to do so. After installing the npm modules, you should be able to invoke `npm start:site` to get running locally.

Note, it does require a valid `results.json` file at the root of the project directory, so you'll have to drop one in or produce one via `npm start`.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [5.0.2](https://github.com/m3fawner/webperf-comparison/compare/v5.0.1...v5.0.2) (2021-12-15)

### [5.0.1](https://github.com/m3fawner/webperf-comparison/compare/v5.0.0...v5.0.1) (2021-09-14)

## [5.0.0](https://github.com/m3fawner/webperf-comparison/compare/v4.0.0...v5.0.0) (2021-09-14)


### ⚠ BREAKING CHANGES

* Log prompt answers to the results.json, consume them in the site to provide context to previous run values

#### Breaking change
* `hosts` became a nested property of promptAnswers, breaking previous results.json outputs

### Features

* Log prompt answers to the results.json, consume them in the site to provide context to previous run values ([274b91f](https://github.com/m3fawner/webperf-comparison/commit/274b91f87b37538f423f29d5af436ce60b6793f6))

## [4.0.0](https://github.com/m3fawner/webperf-comparison/compare/v3.1.0...v4.0.0) (2021-09-13)


### ⚠ BREAKING CHANGES

* **cli:** Allow a list of hosts, instead of assuming original and comparison. This allows for single runs, as well as more than 2 hosts.

#### Breaking change
* originalURL is deprecated
* comparisonURL is deprecated
* *New property* `hosts`: Comma delimited list of hostnames to run lighthouse against. Allows for single host, or many. Replaces originalURL/comparisonURL

### Features

* **cli:** Allow a list of hosts, instead of assuming original and comparison. This allows for single runs, as well as more than 2 hosts. ([586ad87](https://github.com/m3fawner/webperf-comparison/commit/586ad879a8bd5d6434e316cee07eb189b7b17131))

## [3.1.0](https://github.com/m3fawner/webperf-comparison/compare/v1.4.0...v3.1.0) (2021-09-08)


### Features

* **cli/site:** Add first input delay measurements ([935478a](https://github.com/m3fawner/webperf-comparison/commit/935478aca13892db9abd600f00454b5d91df7d13))
* **cli:** Throttling of network & CPU added, simulated via lighthouse ([3652f3e](https://github.com/m3fawner/webperf-comparison/commit/3652f3eaee4288bf1d51fff5ae2944946dd83338))

## [3.0.0](https://github.com/m3fawner/webperf-comparison/compare/v1.4.0...v3.0.0) (2021-09-03)


### ⚠ BREAKING CHANGES

* **cli:** Replace usage of the new keyword throughout code and prompts.

### Features

* **cli/site:** Add first input delay measurements ([ddd61e1](https://github.com/m3fawner/webperf-comparison/commit/ddd61e1b7eef0f494d3bb9deb9d030f1f5068400))


* **cli:** Replace usage of the new keyword throughout code and prompts. ([d936a46](https://github.com/m3fawner/webperf-comparison/commit/d936a469fc9f4c8f59da85b74cf06d2c22b4bf8b))

## [2.0.0](https://github.com/m3fawner/webperf-comparison/compare/v1.4.0...v2.0.0) (2021-08-03)


### ⚠ BREAKING CHANGES

**Impact** Previous `results.json` files will no longer be loadable by `--usePrevious`, `npm run start:previous` or by `npm run start:site`

* **cli:** Replace usage of the new keyword throughout code and prompts. ([d936a46](https://github.com/m3fawner/webperf-comparison/commit/d936a469fc9f4c8f59da85b74cf06d2c22b4bf8b))

## [1.4.0](https://github.com/m3fawner/webperf-comparison/compare/v1.3.0...v1.4.0) (2021-08-03)


### Features

* **cli:** Run every other instead of full group to decrease likelihood of outside factors causing variance ([ce2e3fc](https://github.com/m3fawner/webperf-comparison/commit/ce2e3fc2e527489724a35e8a08af05be6fa74582))
* **site:** Add box & whisker plots for metric groups ([d108c1f](https://github.com/m3fawner/webperf-comparison/commit/d108c1f27ad13cdb9435fa144499586377498c70))
* **site:** Regroup standard deviation table to be an accordion per metric ([28660ff](https://github.com/m3fawner/webperf-comparison/commit/28660ff179483d8995d1e7140575e59c58e40de3))

## [1.3.0](https://github.com/m3fawner/webperf-comparison/compare/v1.2.0...v1.3.0) (2021-07-29)


### Features

* **site:** I couldn't live with myself after writing the readme update. React/Chakra-fied ([cdb98bf](https://github.com/m3fawner/webperf-comparison/commit/cdb98bfe879417b6170de20d58f0c11f09b52cc1))

## [1.2.0](https://github.com/m3fawner/webperf-comparison/compare/v1.1.0...v1.2.0) (2021-07-29)


### Features

* **printout:** Enhance the print out table to use the new and comparison URLs instead of new/comp to be explicit ([6d7239a](https://github.com/m3fawner/webperf-comparison/commit/6d7239ac17a946993c7e04e6f3d53b879f2ee7f0))
* **site:** Produce a quick and dirty page for consuming the results in more depth, presently standard variation ([690cd6c](https://github.com/m3fawner/webperf-comparison/commit/690cd6c4ca0f248dd09a746dc2ff55fe182ed1cf))

## [1.1.0](https://github.com/m3fawner/webperf-comparison/compare/v1.0.0...v1.1.0) (2021-07-20)


### Features

* Round output more effectively to account for larger numbers ([478d5b4](https://github.com/m3fawner/webperf-comparison/commit/478d5b499dd6e4570b7037bfb50e364542621e0b))

## 1.0.0 (2021-07-20)


### Features

* Initial implementation ([a290a9a](https://github.com/m3fawner/webperf-comparison/commit/a290a9afbc5959fc0e728b8eeb91c9ee174740eb))


constants.js
const METRIC_LOOKUP = {
  FCP: 'fcp',
  LCP: 'lcp',
  INT: 'int',
  SPD: 'spd',
  TBT: 'tbt',
  CLS: 'cls',
  FID: 'fid',
};
const METRIC_LABELS = {
  ROUTE: 'Route',
  HOST: 'Host',
  [METRIC_LOOKUP.FCP]: 'FCP',
  [METRIC_LOOKUP.LCP]: 'LCP',
  [METRIC_LOOKUP.FID]: 'FID',
  [METRIC_LOOKUP.INT]: 'Interactive',
  [METRIC_LOOKUP.SPD]: 'Speed',
  [METRIC_LOOKUP.TBT]: 'Blocking Time',
  [METRIC_LOOKUP.CLS]: 'CLS',
};
const METRIC_LABELS_LONG = {
  [METRIC_LOOKUP.FCP]: 'First Contentful Paint',
  [METRIC_LOOKUP.LCP]: 'Largest Contentful Paint',
  [METRIC_LOOKUP.FID]: 'First Input Delay',
  [METRIC_LOOKUP.INT]: 'Time to Interactive',
  [METRIC_LOOKUP.SPD]: 'Speed Index',
  [METRIC_LOOKUP.TBT]: 'Total Blocking Time',
  [METRIC_LOOKUP.CLS]: 'Cumulative Layout Shift',
};
const METRIC_HINTS = {
  [METRIC_LOOKUP.FCP]: 'First Contentful Paint marks the time at which the first text or image is painted.First Contentful Paint marks the time at which the first text or image is painted.',
  [METRIC_LOOKUP.LCP]: 'Largest Contentful Paint marks the time at which the largest text or image is painted.',
  [METRIC_LOOKUP.FID]: 'First Input Delay measures interactivity. To provide a good user experience, pages should have a FID of 100 milliseconds or less.',
  [METRIC_LOOKUP.INT]: 'Time to interactive is the amount of time it takes for the page to become fully interactive',
  [METRIC_LOOKUP.SPD]: 'Speed Index shows how quickly the contents of a page are visibly populated.',
  [METRIC_LOOKUP.TBT]: 'Sum of all time periods between FCP and Time to Interactive, when task length exceeded 50ms, expressed in milliseconds.',
  [METRIC_LOOKUP.CLS]: 'Cumulative Layout Shift measures the movement of visible elements within the viewport.',
};
const METRICS = {
  [METRIC_LOOKUP.FCP]:
  'firstContentfulPaint',
  [METRIC_LOOKUP.LCP]: 'largestContentfulPaint',
  [METRIC_LOOKUP.FID]: 'maxPotentialFID',
  [METRIC_LOOKUP.INT]: 'interactive',
  [METRIC_LOOKUP.SPD]: 'speedIndex',
  [METRIC_LOOKUP.TBT]: 'totalBlockingTime',
  [METRIC_LOOKUP.CLS]: 'cumulativeLayoutShift',
};
const NETWORK_THROTTLES_LOOKUP = {
  THREE_G_SLOW: '3gslow',
  THREE_G: '3g',
  THREE_G_FAST: '3gfast',
  CABLE: 'cable',
  FOUR_G: '4g',
};
const NETWORK_THROTTLES = {
  [NETWORK_THROTTLES_LOOKUP.THREE_G_SLOW]: '3gslow',
  [NETWORK_THROTTLES_LOOKUP.THREE_G]: '3g',
  [NETWORK_THROTTLES_LOOKUP.THREE_G_FAST]: '3gfast',
  [NETWORK_THROTTLES_LOOKUP.CABLE]: 'cable',
  [NETWORK_THROTTLES_LOOKUP.FOUR_G]: '4g',
};
const NETWORK_THROTTLES_OPTIONS = {
  [NETWORK_THROTTLES_LOOKUP.THREE_G_SLOW]: {
    uploadThroughputKbps: 400,
    downloadThroughputKbps: 400,
    rttMs: 200,
  },
  [NETWORK_THROTTLES_LOOKUP.THREE_G]: {
    uploadThroughputKbps: 768,
    downloadThroughputKbps: 1600,
    rttMs: 150,
  },
  [NETWORK_THROTTLES_LOOKUP.THREE_G_FAST]: {
    uploadThroughputKbps: 768,
    downloadThroughputKbps: 1600,
    rttMs: 75,
  },
  [NETWORK_THROTTLES_LOOKUP.CABLE]: {
    uploadThroughputKbps: 1000,
    downloadThroughputKbps: 5000,
    rttMs: 14,
  },
  [NETWORK_THROTTLES_LOOKUP.FOUR_G]: {
    uploadThroughputKbps: 9000,
    downloadThroughputKbps: 9000,
    rttMs: 85,
  },
};
const CPU_THROTTLES_LOOKUP = {
  SLOW_MOBILE_CPU: 'slow_mobile',
  MOBILE_CPU: 'mobile',
  DESKTOP: 'desktop',
};
const CPU_THROTTLES = {
  [CPU_THROTTLES_LOOKUP.SLOW_MOBILE_CPU]: 6,
  [CPU_THROTTLES_LOOKUP.MOBILE_CPU]: 4,
  [CPU_THROTTLES_LOOKUP.DESKTOP]: 1,
};
const CPU_THROTTLES_DESCRIPTIONS = {
  [CPU_THROTTLES_LOOKUP.SLOW_MOBILE_CPU]: 'Slow mobile phone',
  [CPU_THROTTLES_LOOKUP.MOBILE_CPU]: 'Mobile phone',
  [CPU_THROTTLES_LOOKUP.DESKTOP]: 'Desktop',
};

const PROMPT_KEY_LOOKUP = {
  CPU: 'cpu',
  HOSTS: 'hosts',
  LOAD_SITE: 'loadSite',
  NETWORK: 'network',
  ROUTES: 'routes',
  RUNS: 'runs',
};
const PROMPT_DESCRIPTIONS = {
  [PROMPT_KEY_LOOKUP.CPU]: 'Which CPU throttling option?',
  [PROMPT_KEY_LOOKUP.HOSTS]: 'What are the hosts that you wish to test?',
  [PROMPT_KEY_LOOKUP.LOAD_SITE]: 'Would you like to view the full results via your browser?',
  [PROMPT_KEY_LOOKUP.NETWORK]: 'Which network throttling option?',
  [PROMPT_KEY_LOOKUP.ROUTES]: 'What routes (paths) would you like to test between the host(s)?',
  [PROMPT_KEY_LOOKUP.RUNS]: 'How many times, per experience, would you like to run the lighthouse runner?',
};

module.exports = {
  METRICS,
  METRIC_LABELS,
  METRIC_LOOKUP,
  METRIC_HINTS,
  METRIC_LABELS_LONG,
  NETWORK_THROTTLES,
  NETWORK_THROTTLES_OPTIONS,
  CPU_THROTTLES_LOOKUP,
  CPU_THROTTLES,
  CPU_THROTTLES_DESCRIPTIONS,
  PROMPT_KEY_LOOKUP,
  PROMPT_DESCRIPTIONS,
};
